### PR TITLE
feat(number-theory): compute exact Ramanujan sums

### LINKED ISSUE #2314: [Operation]: Compute exact Ramanujan sums from divisor–Möbius data
## Problem

The Erdős 1002 formalization repeatedly needs the exact integer Ramanujan sum

\[
c_q(n)=\sum_{\substack{0\le a<q\\(a,q)=1}} e^{2\pi ian/q},
\]

including signed frequencies, the modulus-zero convention, and finite-period
orthogonality. Jacobian currently exposes the ingredients (`gcd`, positive
divisors, and the Möbius function), but not the named exact arithmetic value. A
caller can reconstruct it from

\[
c_q(n)=\sum_{d\mid \gcd(q,|n|)} d\,\mu(q/d),
\]

only by orchestrating a variable number of scalar calls and carrying the source
relation itself.

This is a narrowly scoped operation proposal, not a Fourier-series solver. There
is a real admission tradeoff: if the project considers this formula a cheap
projection of retained divisor/factorization data, the same source is still
useful as a cross-operation fixture for #2311 rather than a new catalog entry.

## Source demand

The formal development defines the reduced-residue character sum, proves its
modulus-zero and signed-frequency conventions, and derives the exact
divisor–Möbius formula:

- [`RamanujanSums.lean`](https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/1002/lean/Erdos1002/RamanujanSums.lean)
- [`RamanujanIdentities.lean`](https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/1002/lean/Erdos1002/RamanujanIdentities.lean#L220-L310)
- downstream use in [`MobiusCollapse.lean`](https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/1002/lean/Erdos1002/MobiusCollapse.lean) and [`MobiusCotangent.lean`](https://github.com/ShouqiaoW/erdos/blob/d28713ac8245ca86a686b8c67370a8d19d81b242/1002/lean/Erdos1002/MobiusCotangent.lean)

A targeted issue search found no existing owner. #960 concerns exact
finite-Abelian spectral-pair verification; #1743 concerns modular forms and
explicitly avoids theorem-specific Ramanujan endpoints.

## Proposed operation

`number_theory.ramanujan_sum.compute`

Input:

- bounded modulus `q >= 0`;
- bounded signed frequency `n`.

Return:

- the exact integer `c_q(n)`;
- source-bound `q` and `n`;
- optionally the canonical divisor contribution ledger
  `(d, mu(q/d), d*mu(q/d))` when that does not duplicate a retained
  factorization value.

Conventions should be explicit:

- `c_0(n)=0` (empty reduced-residue sum);
- `c_q(-n)=c_q(n)`;
- `c_q(0)=phi(q)` for positive `q`;
- positive-modulus values use the divisor–Möbius formula, not floating roots
  of unity.

## Bounds and implementation

Validate before work:

- integer digit bounds for `q` and `n`;
- factorization/divisor count and factor-certificate policy inherited from the
  number-theory owner;
- contribution-row and aggregate-result limits if a ledger is returned.

A direct exact formula over the retained factorization/divisor representation is
sufficient. There must be no approximate complex fallback and no enumeration of
all reduced residues when the factorization formula fits the accepted domain.

## Defining tests

Known answers:

- `c_0(n)=0`;
- `c_1(n)=1`;
- `c_2(n)=(-1)^n`;
- `c_3(n)=2` when `3 | n`, otherwise `-1`;
- `c_4(0)=2`, `c_4(2)=-2`, and `c_4(n)=0` for odd `n`;
- `c_5(0)=4`, otherwise `-1` modulo 5.

Properties:

- evenness in `n`;
- period `q` for positive `q`;
- `c_q(0)=phi(q)`;
- exact equality with the divisor–Möbius formula;
- multiplicativity in coprime moduli;
- complete-period orthogonality, including `(q,r)=(2,2)->2`, `(2,3)->0`,
  `(3,3)->6`, and `(3,4)->0`;
- source mutation of either `q` or `n` invalidates an authored result;
- requests exactly at and above integer/factorization/contribution bounds.

## Non-goals

- symbolic or analytic Fourier transforms;
- principal-value/cotangent evaluation;
- Ramanujan expansions of arbitrary functions;
- modular-form Ramanujan conjectures;
- proving the Erdős 1002 Möbius-collapse or limit theorem.

## Acceptance criteria

- [ ] The operation returns the exact integer value for the declared conventions.
- [ ] The result cannot detach from `(q,n)`.
- [ ] Work and result growth are admitted before factorization/divisor expansion.
- [ ] No floating complex arithmetic participates in the result.
- [ ] Known-answer, divisor-formula, periodicity, orthogonality, mutation, and
      boundary tests pass.
- [ ] If owner review classifies this as ordinary composition instead, the
      fixture set is added to #2311 and no redundant public operation is
      introduced.

### PR DESCRIPTION
**Pull request by morluto**
Closes #2314.

### Summary

Add `number_theory.ramanujan_sum.compute` for the exact integer Ramanujan sum `c_q(n)` at a bounded nonnegative modulus and signed frequency. The operation preserves the formalization's explicit `c_0(n) = 0` convention and returns a result bound to both `q` and `n`.

The kernel uses SymPy 1.14 for the bounded prime-power factorization of `q`, then evaluates the exact multiplicative prime-power formula with integer arithmetic. It does not enumerate reduced residues or use floating roots of unity. A native `jacobian.math.number_theory.ramanujan_sum` function exposes the same mathematical map directly.

The result intentionally contains only the source and exact scalar value. A divisor contribution ledger would duplicate factorization data without improving downstream composition; result validation instead replays the defining computation inside the same bounded envelope.

### Public operation contract

- **Semantic mathematical domain and postcondition:** for `q >= 0` and `n in Z`, return the classical exact Ramanujan sum `c_q(n)`, with the empty reduced-residue convention `c_0(n) = 0`.
- **Canonical public value type:** the native value is a Python integer; the wire request/result use canonical decimal integer strings. `RamanujanSumResult` is source-bound to `(q, n)`.
- **Source representation:** two materialized scalar integers.
- **Expansion performed by the kernel and its pre-execution bound:** positive `q` is factored once per computation or replay. A 12-digit modulus has at most 11 distinct prime factors and total prime multiplicity at most 39; no divisor or residue family is expanded.
- **Representation-sensitive complexity:** accepting an already-factored modulus could avoid factorization but would create a parallel source representation. This contract keeps one canonical scalar modulus and a conservative factorization envelope.
- **Admitted request envelope and controlling quantities:** `q` is a canonical nonnegative integer with at most 12 digits, matching the existing in-process factorization envelope; `n` is a canonical signed integer with at most 256 characters and participates only in modular reductions.
- **Producer/consumer closure:** no operation currently consumes a Ramanujan-sum-specific value. The native integer is immediately reusable by ordinary exact arithmetic; the wire result retains all source context needed to replay the relation.
- **Degenerate inputs:** `q = 0` returns `0`, `q = 1` returns `1`, and `n = 0` returns `phi(q)` for positive `q`.
- **Parent/ring/field identity:** inputs and output are integers; no coercion to a complex or floating field occurs.
- **Deterministic work bound:** bounded SymPy factorization of `q`, followed by one constant-size prime-power branch per factor. Result validation repeats the same bounded computation to reject forged source/value combinations.
- **Maximum intermediate growth:** every prime power is at most `q`; modular operands use the admitted frequency length; the accumulated exact product has magnitude at most `q`.
- **Exact result-size bound:** `|c_q(n)| <= phi(q) <= q`, so an accepted positive modulus produces at most 12 value digits plus a sign.
- **Backend and supported version:** pinned SymPy 1.14.0, used privately for `factorint`.
- **Backend input domain:** strictly positive Python integers within the 12-digit request bound; `q = 0` is handled before the backend call.
- **Conversion/coercion behavior:** canonical decimal strings are parsed to Python integers once at the operation boundary; the backend result is converted to exact Python integers only.
- **Result type:** `RamanujanSumResult(modulus, frequency, value)`.
- **Reconstruction or defining invariant:** the value agrees with the prime-power formula, cross-checked against `sum_{d | gcd(q, |n|)} d * mu(q/d)`; result validation replays that relation.
- **Typed execution failures:** malformed, negative, or over-budget moduli and malformed or over-budget frequencies fail request validation. Accepted requests have no approximate, incomplete, or unknown result mode.
- **Property and boundary tests:** known values, divisor–Möbius agreement, evenness, periodicity, coprime-modulus multiplicativity, complete-period orthogonality, source/result mutation rejection, the admitted boundary, and immediately rejected bounds.

### Library choice

SymPy already owns the maintained exact factorization kernel used throughout this number-theory domain, but it does not expose a Ramanujan-sum routine. Keeping factorization private and implementing the short prime-power identity locally is smaller and more exact than introducing another dependency or an approximate complex path.

### Validation

- `make test-math TESTS=tests/math/number_theory/test_ramanujan_sums.py` — 20 passed.
- `make check` — Ruff and mypy passed; 2,669 tests passed.

### CODE REVIEW

File: src/jacobian/math/number_theory/_ramanujan_sum.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    """An exact Ramanujan sum bound to its modulus and frequency."""
+
+    modulus: RamanujanModulus
+    frequency: BoundedInteger
+    value: BoundedInteger
+
+    @model_validator(mode="after")
+    def bind_value_to_source(self) -> Self:
+        expected = ramanujan_sum(int(self.modulus), int(self.frequency))
+        if int(self.value) != expected:
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/_ramanujan_sum.py:84]: Fixed in 698c28e33: RamanujanSumResult.value now uses the domain-owned RamanujanSumInteger grammar ^(?:0|-?[1-9][0-9]*)$ (the shared canonical signed-decimal pattern), rejecting "-0", "+0", leading zeros, whitespace and underscores before binding; the producer emits str(int), so zero sums serialize canonically as "0" (modulus=4, frequency=3 regression). Lint/typecheck/number-theory lane/catalog green.

**Comment by morluto** [comment]
@codex

**Comment by morluto** [comment]
@codex

File: src/jacobian/math/number_theory/_ramanujan_sum.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    """One bounded modulus and signed frequency for an exact Ramanujan sum."""
+
+    modulus: RamanujanModulus = Field(
+        description=(
+            "Canonical nonnegative integer modulus with at most 12 digits; "
+            "zero denotes the empty reduced-residue sum."
+        )
+    )
+    frequency: BoundedInteger = Field(
+        description="Canonical signed integer frequency with at most 256 characters."
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/_ramanujan_sum.py]: Root cause: the frequency request field originally reused BoundedInteger, whose grammar admits "-0", so mathematically identical zero frequencies could enter with distinct serialized identities and be echoed into the source-bound result.

Fix (completed in bf741cf7d, now finished on top of it): every integer field on the request and result binds through a single constraint. The operation-specific 256-character execution bound is applied as Field(max_length=_MAX_INTEGER_LENGTH) on top of the owner type, never by restating its grammar, so Ramanujan frequencies and sums cannot diverge from other canonical-integer producers or consumers.

Validation: tests/math/number_theory/test_ramanujan_sums.py passes (64 passed), including test_negative_zero_frequency_is_rejected_before_source_binding, test_request_rejects_noncanonical_frequency_encodings, and new test_frequency_grammar_is_owned_by_canonical_integer; ruff and mypy clean.

File: src/jacobian/math/number_theory/_ramanujan_sum.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+# optional minus sign on a nonzero leading digit.  Unlike
+# ``BoundedInteger``, whose grammar also admits "-0", this pattern cannot
+# bind negative zero, so every accepted string is the unique canonical
+# decimal form of its integer and mathematically equal inputs share one
+# serialized identity across request and result.
+RamanujanSumInteger = Annotated[
+    str,
+    StringConstraints(
+        pattern=r"^(?:0|-?[1-9][0-9]*)$",
+        max_length=_MAX_INTEGER_LENGTH,
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/_ramanujan_sum.py]: Root cause: RamanujanSumInteger restated the canonical signed-decimal pattern locally, duplicating a grammar already owned by jacobian._exact.CanonicalInteger; matching today is fragile because any future normalization or schema change to the owner type would silently leave Ramanujan frequencies and results behind.

Fix: RamanujanSumInteger is now Annotated[CanonicalInteger, Field(max_length=_MAX_INTEGER_LENGTH)]. The owner type supplies the grammar (including negative-zero rejection); the operation-specific 256-character execution bound is applied separately as a field constraint. The JSON schema still exposes pattern ^(?:0|-?[1-9][0-9]*)$ with maxLength 256.

Validation: tests/math/number_theory/test_ramanujan_sums.py passes (64 passed), including new test_frequency_grammar_is_owned_by_canonical_integer pinning request acceptance to TypeAdapter(CanonicalInteger), plus the existing boundary test test_ramanujan_sum_request_bounds_factorization_and_frequency_work (256 accepted, 257 rejected). Full tests/math/number_theory lane passes (95 passed); ruff and mypy clean.

File: src/jacobian/math/number_theory/_ramanujan_sum.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+# This matches the established in-process factorization envelope.  SymPy
+# factors the modulus once; the frequency only participates in bounded modular
+# reductions.  The exact result has absolute value at most the modulus.
+_MAX_MODULUS_DIGITS = 12
+
+RamanujanModulus = Annotated[
+    str,
+    StringConstraints(
+        pattern=r"^(?:0|[1-9][0-9]*)$",
+        max_length=_MAX_MODULUS_DIGITS,
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/_ramanujan_sum.py]: Root cause: RamanujanModulus restated the canonical decimal grammar with its own StringConstraints pattern instead of binding through jacobian._exact.CanonicalInteger, so the modulus was a parallel encoding outside the owner-conformance test and could silently diverge from other canonical-integer producers/consumers if the owner grammar changes.

Fix: RamanujanModulus is now Annotated[CanonicalInteger, Field(max_length=12)] - the owner grammar with only the operation-specific 12-character execution bound applied on top, matching how the frequency/result fields already bind. Nonnegativity is enforced separately by a require_nonnegative_modulus model validator on both RamanujanSumRequest and RamanujanSumResult (the accepted request domain is unchanged: "-1" was rejected before and still is). The shared _MAX_MODULUS_DIGITS bound now lives in the ramanujan_sums kernel next to the factorization-work admission that uses it.

Tests: test_modulus_grammar_is_owned_by_canonical_integer extends the owner-conformance check to the modulus; test_owner_grammar_admits_negative_moduli_the_operation_rejects proves the owner grammar admits negative moduli that operation validation rejects (request and forged result). Validated: tests/math/number_theory lane, make test-catalog, make test-dispatch, ruff, mypy.

File: src/jacobian/math/number_theory/ramanujan_sums.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    """Return the exact Ramanujan sum ``c_modulus(frequency)``.
+
+    The zero modulus denotes the empty reduced-residue sum.  Positive moduli
+    are evaluated multiplicatively from their prime-power factorization, so
+    no approximate roots of unity or reduced-residue enumeration is involved.
+    """
+
+    q = modulus.__index__()
+    n = frequency.__index__()
+    if q < 0:
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/ramanujan_sums.py:33]: Root cause: the factorization-work envelope existed only as a wire constraint (12-character modulus) on RamanujanSumRequest. The natively exported ramanujan_sum() bypasses that request model, so a direct caller could hand an arbitrarily large modulus straight to SymPy factorint with unbounded work.

Fix: the kernel now enforces the same named budget before backend expansion: ramanujan_sums.ramanujan_sum owns _MAX_MODULUS_DIGITS = 12 and raises ValueError("a Ramanujan-sum modulus must have at most 12 decimal digits") when q >= 10**12, checked after the zero-modulus early return and before the lazy factorint import. The operation module imports this constant for its Field(max_length=...), so both public entry points share one bound and neither can reach the backend outside the declared envelope.

Tests: test_native_ramanujan_sum_bounds_factorization_work rejects a 13-digit modulus via the native entry point and proves the 12-digit boundary case stays admitted with a known answer (c_549755813888(274877906944) = -274877906944). Validated: tests/math/number_theory lane, make test-catalog, make test-dispatch, ruff, mypy.

File: src/jacobian/math/number_theory/ramanujan_sums.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    The zero modulus denotes the empty reduced-residue sum.  Positive moduli
+    are evaluated multiplicatively from their prime-power factorization, so
+    no approximate roots of unity or reduced-residue enumeration is involved.
+    A positive modulus must have at most 12 decimal digits so that the
+    factorization work stays bounded for every entry point.
+    """
+
+    q = modulus.__index__()
+    n = frequency.__index__()
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/number_theory/ramanujan_sums.py:32]: Root cause: the native ramanujan_sum accepted any SupportsIndex frequency while only the wire model bounded it to 256 characters, so modulus>1 calls performed unbounded % reductions on arbitrarily large integers. Fix: after the nonnegativity check and before factorint or any modular reduction, the frequency magnitude is rejected unless it fits exactly the same canonical 256-character envelope (positive up to 10^256-1, negative down to -(10^255-1)), checked via magnitudes without materializing digits; the zero-modulus shortcut now also sits behind the bound for exact wire parity. Changed src/jacobian/math/number_theory/ramanujan_sums.py with adversarial boundary tests in tests/math/number_theory/test_ramanujan_sums.py covering rejection above the envelope at both signs, acceptance at the boundaries, and agreement between the native path and RamanujanSumRequest execution. Validated with make check (3876 passed) and make test-math TESTS="tests/math/number_theory" (164 passed).
